File: Week2/Exercise1.js
const readline=require('readline');
const r1=readline.createInterface({input:process.stdin,output:process.stdout});
const r2=readline.createInterface({input:process.stdin,output:process.stdout});

let array=[];
r1.question("Enter the elements of the array separated by spaces:",ans=> {
    array=ans.split(" ").map(Number);
    console.log("The array you entered is:",array);
    
let array2=[];
for(let i=0;i<array.length;i++){
    array2[i]=array[i];
}

//Insertion Sort 
for (i=1;i<array2.length;i++)
    {  
     let j=i;
    while (j > 0 && array2[j] < array2[j - 1]) {
            let temp = array2[j];
            array2[j] = array2[j - 1];
            array2[j - 1] = temp;
            j--;
        }
    }

console.log("The Insertion Sort array is:",array2);

//Bubble Sort
for(let i=0;i<array.length;i++)
    {
        for(let j=0;j<array.length-1;j++)
            {
            if(array[j]>array[j+1])
                {
                    let temp=array[j];
                    array[j]=array[j+1];
                    array[j+1]=temp;
                }
            }
    }
console.log("The Bubble sort array is:",array);

//Merge Sort
let arr=[];
for(let i=0;i<array.length;i++){
    arr[i]=array[i];
}
function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    let mid = Math.floor(arr.length / 2);
    let left = arr.slice(0, mid);
    let right = arr.slice(mid);

    return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
    let sorted = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            sorted.push(left[i]);
            i++;
        } else {
            sorted.push(right[j]);
            j++;
        }
    }

    // Add remaining elements
    return sorted.concat(left.slice(i)).concat(right.slice(j));
}
let result = mergeSort(array);
console.log("Merge sort array is :", result);

//Linear Search
r1.question("Enter the element to be searched:",ans2=>{
    let target=Number(ans2);
    let found=false;
    for(let i=0;i<array.length;i++){
        if(array[i]===target){
            found=true;
            console.log("Element found at index:",i);
            break;
        }   
        else{
            found=false;
        }
    }
    if(!found){
        console.log("Element not found in the array");
    }
    
//Binary Search
 r1.question("Enter the element to be searched using Binary Search:",ans3=>{
    let target=Number(ans3);
    let left=0;
    let right=array.length-1;
    let found=false;
    while(left<=right){
        let mid=Math.floor((left+right)/2);
        if(array[mid]===target){
            found=true;
            console.log("Element found at index:",mid);
            break;
        }
        else if(array[mid]<target){
            left=mid+1;
        }
        else{
            right=mid-1;
        }
    }
    if(!found){
        console.log("Element not found in the array");
    }
r1.close();
});
});
});

File: Week2/Exercise1.py
array=[]
array2=[]
count=int(input("Enter number of elements: "))
for i in range(count):
    element=int(input("Enter element: "))
    array.append(element)

print("Original array:", array)
array2=array.copy()
# insertion sort
for k in range(1,len(array2)):
    j=k
    while(j>0 and array2[j] < array2[j-1]):
            temp=array2[j]
            array2[j]=array2[j-1]
            array2[j-1]=temp
            j=j-1

print("Insertion Sort Array:", array2)

# bubble sort
for i in range(len(array)):
    j=0
    while(j<=len(array)-2):
        if(array[j]>array[j+1]):
            temp=array[j]
            array[j]=array[j+1]
            array[j+1]=temp
        j=j+1
print("Bubble Sort Array:", array)

# merge sort
array3=array.copy()
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]

        merge_sort(L)
        merge_sort(R)

        i = j = k = 0

        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1
merge_sort(array3)
print("Merge Sort Array:", array3)


# Linear Search
target=int(input("Enter element to search (Linear Search): "))
for i in range(len(array)):
    if(array[i]==target):
        print("Element found at index (Linear Search):",i)
        break
    else:
        print("Element not found (Linear Search)")

# Binary Search
target=int(input("Enter element to search (Binary Search): "))
left=0
right=len(array3)-1
found=False
while left<=right:
    mid=(left+right)//2
    if(array3[mid]==target):
        print("Element found at index (Binary Search):",mid)
        found=True
        break
    elif(array3[mid]<target):
        left=mid+1
    else:
        right=mid-1
    
if not found:
    print("Element not found (Binary Search)")
